ledger-tool/src/bigtable.rs

```diff
@@ -164,6 +172,148 @@ async fn entries(
     Ok(())
 }
 
+struct ShredConfig {
+    shred_version: u16,
+    num_hashes_per_tick: u64,
+    num_ticks_per_slot: u64,
+    allow_mock_poh: bool,
+}
+
+async fn shreds(
+    blockstore: Arc<Blockstore>,
+    starting_slot: Slot,
+    ending_slot: Slot,
+    shred_config: ShredConfig,
+    config: solana_storage_bigtable::LedgerStorageConfig,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let bigtable = solana_storage_bigtable::LedgerStorage::new_with_config(config)
+        .await
+        .map_err(|err| format!("Failed to connect to storage: {err:?}"))?;
+
+    // Make the range inclusive of both starting and ending slot
+    let limit = (ending_slot - starting_slot + 1) as usize;
+    let mut slots = bigtable.get_confirmed_blocks(starting_slot, limit).await?;
+    slots.retain(|&slot| slot <= ending_slot);
+
+    let keypair = Keypair::from_bytes(&[0; 64])?;
+    let ShredConfig {
+        shred_version,
+        num_hashes_per_tick,
+        num_ticks_per_slot,
+        allow_mock_poh,
+    } = shred_config;
+
+    for slot in slots.iter() {
+        let block = bigtable.get_confirmed_block(*slot).await?;
+        let entry_summaries = match bigtable.get_entries(*slot).await {
+            Ok(summaries) => Some(summaries),
+            Err(err) => {
+                let err_msg = format!("Failed to get PoH entries for {slot}: {err}");
+
+                if allow_mock_poh {
+                    warn!("{err_msg}. Will create mock PoH entries instead.");
+                } else {
+                    return Err(format!(
+                        "{err_msg}. Try passing --allow-mock-poh to allow \
+                        creation of shreds with mocked PoH entries"
+                    ))?;
+                }
+                None
+            }
+        };
+
+        let entries = match entry_summaries {
+            Some(entry_summaries) => entry_summaries
+                .map(|entry_summary| {
+                    let num_hashes = entry_summary.num_hashes;
+                    let hash = entry_summary.hash;
+                    let transactions = block.transactions[entry_summary.starting_transaction_index
+                        ..entry_summary.starting_transaction_index
+                            + entry_summary.num_transactions as usize]
```

Comment:
As a nicety, how about we check that `block.transactions.len() - 1` equals the last entry_summary's `starting_transaction_index + num_transactions` before this so we return a more descriptive error if something was recorded badly?

Reply:
> To confirm, you're thinking about a check to ensure we don't do an out-of-bounds array access right ?

Yes, exactly. Sure, that check works! I was just trying to shortcut having to do it every iteration, but it's not too heavy.
Thanks, I knew I should have double-checked for off-by-one.

Reply:
I ended up using `Vec::get()` which does the bounds check internally in https://github.com/anza-xyz/agave/pull/1638/commits/6a075105588a3f125151728fc784891ff39814cd